sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
-        std::tie(_current_range, _current_sstables, _current_next_position) = _impl->select(pos);
+        std::vector<shared_sstable> _current_versioned_sstables;
+        std::tie(_current_range, _current_versioned_sstables, _current_next_position) = _impl->select(pos);
+        _current_sstables = boost::copy_range<std::vector<shared_sstable>>(_current_versioned_sstables
+                             | boost::adaptors::filtered([this] (shared_sstable sst) { return _sstl->contains(sst); })
+                             | boost::adaptors::transformed([] (shared_sstable sst) { return sst; }));
         _current_range_view = _current_range->transform([] (const dht::ring_position& rp) { return dht::ring_position_view(rp); });
     }
     return {_current_sstables, _current_next_position};
 }
 
-sstable_set::incremental_selector
-sstable_set::make_incremental_selector() const {
-    return incremental_selector(_impl->make_incremental_selector(), *_schema);
+sstables::sstable_set::incremental_selector sstable_set::make_incremental_selector() const {
+    return incremental_selector(_all->version().make_incremental_selector(), *_all->version().get_schema(), _all);
+}
+
+sstable_set_version::~sstable_set_version() {
+    for (auto&& added_sst : _added) {
+        _versioned_data->remove(added_sst);
+    }
+    if (_prev) {
+        _prev->_next.erase(this);
+    }
+}
+
+sstable_set_version::sstable_set_version(std::unique_ptr<sstable_set_impl> impl, schema_ptr schema) 
+    : _versioned_data(make_lw_shared<versioned_sstable_set_data>(std::move(impl))) 
+    , _schema(std::move(schema)) {
+}
+
+sstable_set_version::sstable_set_version(sstable_set_version* ver) 
+    : _versioned_data(ver->_versioned_data)
+    , _schema(ver->_schema)
+    , _prev(ver) {
+        _prev->_next.insert(this);
+}
+
+size_t sstable_set_version::sstables_changed() const {
+    return _added.size() + _erased.size();
+}
+
+void sstable_set_version::merge_with_next() {
+    auto next_version = *_next.begin();
+    if (sstables_changed() > next_version->sstables_changed()) {
+        std::swap(_added, next_version->_added);
+        std::swap(_erased, next_version->_erased);
+    }
+    for (auto&& sst : _added) {
+        if (!next_version->_erased.contains(sst)) {
+            next_version->_added.insert(sst);
+            _versioned_data->insert(sst);
+        } else {
+            // the sstable will be removed in the destructor of the merged version
+            next_version->_erased.erase(sst);
+        }
+    }
+    for (auto&& sst : _erased) {
+        if (!next_version->_added.contains(sst)) {
+            next_version->_erased.insert(sst);
+        } else {
+            next_version->_added.erase(sst);
+            _versioned_data->remove(sst);
+        }
+    }
+    if (_prev) {
+        _prev->_next.insert(*_next.begin());
+    }
+    (*_next.begin())->_prev = sstable_set_version_reference(_prev.get()); // destroys this by overwriting the last reference
```

Comment:
Ah, now I understand the "references == 1" thing earlier.

First, such a thing should be documented.

Second, I think that only external references should be counted. This avoids the potential for cycles. Since you every internal reference to `this`  (_next->_prev) you also have a back reference, you know that as long as you have _next then you cannot release the content (but you can propagate it into the children).

Reply:
I'm not sure what does counting only external references help with. If I'm not mistaken, if we were still using the emptiness of `_next` as a condition for releasing a version, we could still in theory have a the same potential cycles. 
I don't see how it could simplify the code either: we would have to add some kind of marker which tells whether a reference is external or not, and some ifs that check whether we should modify the `_reference_count` or not. 
I guess another solution would be to remove sstable_set_version_references from sstable_set_versions altogether and replace them with pointers - in this case we would need to have to do the work of `sstable_set_version_reference`'s destructor in each method that would destroy `_prev` - so in the merge method and in `sstable_set_versions`'s destructor.
We could also remove the class altogether, but that would require adding the same code as in the second option, and in every place where we destroy/reassign `sstable_list`'s `_version`

Reply:
> I'm not sure what does counting only external references help with. 

It's just an intuition. Maybe it's wrong. However, whichever way we choose it must be explained, since there are two ways to understand it.

> If I'm not mistaken, if we were still using the emptiness of `_next` as a condition for releasing a version, we could still in theory have a the same potential cycles.
> I don't see how it could simplify the code either: we would have to add some kind of marker which tells whether a reference is external or not, and some ifs that check whether we should modify the `_reference_count` or not.

The code always knows whether a reference is internal or external. If it's from an sstable_set object, it's external, if it's from an sstable_set_version object, it's internal.

> I guess another solution would be to remove sstable_set_version_references from sstable_set_versions altogether and replace them with pointers - in this case we would need to have to do the work of `sstable_set_version_reference`'s destructor in each method that would destroy `_prev` - so in the merge method and in `sstable_set_versions`'s destructor.

Yes.

> We could also remove the class altogether, but that would require adding the same code as in the second option, and in every place where we destroy/reassign `sstable_list`'s `_version`